integrations/airflow/marquez_airflow/extractors/sql/parser.py

```diff
@@ -85,19 +104,57 @@ def parse(sql: str) -> SqlMeta:
         # We assume only one statement in SQL
         tokens = TokenList(statements[0].tokens)
         log.debug(f"Successfully tokenized sql statement: {tokens}")
+        parser = cls()
+        return parser.recurse(tokens)
 
-        in_tables = []
-        out_tables = []
-
+    def recurse(self, tokens: TokenList):
+        in_tables, out_tables = set(), set()
         idx, token = tokens.token_next_by(t=T.Keyword)
         while token:
-            if _is_in_table(token):
-                idx, in_table = _get_table(tokens, idx)
-                in_tables.append(in_table)
+
+            # Main parser switch
+            if self.is_cte(token):
+                cte_name, cte_intables = self.parse_cte(idx, tokens)
+                for intable in cte_intables:
+                    if intable not in self.ctes:
+                        in_tables.add(intable)
+                if cte_name:
+                    self.ctes.add(cte_name)
+            elif _is_in_table(token):
+                idx, extracted_tables = _get_tables(tokens, idx)
+                for table in extracted_tables:
+                    if table.name not in self.ctes:
+                        in_tables.add(table)
             elif _is_out_table(token):
-                idx, out_table = _get_table(tokens, idx)
-                out_tables.append(out_table)
+                idx, extracted_tables = _get_tables(tokens, idx)
+                out_tables.add(extracted_tables[0])  # assuming only one out_table
 
             idx, token = tokens.token_next_by(t=T.Keyword, idx=idx)
 
-        return SqlMeta(in_tables, out_tables)
+        return SqlMeta(list(in_tables), list(out_tables))
+
+    def is_cte(self, token: T):
+        return token.match(T.Keyword.CTE, values=['WITH'])
+
+    def parse_cte(self, idx, tokens: TokenList):
+        _, group = tokens.token_next(idx, skip_ws=True, skip_cm=True)
+
+        if not group.is_group:
+            return [], None
+
+        # get CTE name
+        offset = 1
+        cte_name = group.token_first(skip_ws=True, skip_cm=True)
+        # handle recursive keyword
+        if cte_name.match(T.Keyword, values=['RECURSIVE']):
```

Comment:
Can you add a test of a `with recursive` querry?

Reply:
Added. Also, fixed, because it was bugged, so good catch :slightly_smiling_face: